### ISSUE TITLE
http3.Server doesn't use tls.Config.GetCertificate

### ISSUE THREAD

**Issue by sotodel** [lucas-clemente/quic-go#2514]
A `http.Server` can pass 2 empty strings to [`ListenAndServeTLS`](https://golang.org/pkg/net/http/#Server.ListenAndServeTLS) in which case the Server will expect and use either of `TLSConfig.Certificates` or `TLSConfig.GetCertificate`

> Filenames containing a certificate and matching private key for the server must be provided if neither the Server's `TLSConfig.Certificates` nor `TLSConfig.GetCertificate` are populated

The `http3.Server` doesn't do this. It immediately tries to load files at the specified locations. Could we change that to rely on the `GetCertificate` supplier function and/or preloaded `Certificates`?

**Comment by marten-seemann** [comment]
> A http.Server can pass 2 empty strings to ListenAndServeTLS in which case the Server will expect and use either of TLSConfig.Certificates or TLSConfig.GetCertificate

Can you point me to the code where this occurs?

I'd be happy to review a PR, if you'd like to contribute some code for this.

**Comment by sotodel** [comment]
This happens [here](https://github.com/golang/go/blob/master/src/net/http/server.go#L2969-L2977). Assuming you create the `http.Server` with a `tls.Config` that sets, for example, the `GetCertificate`, then the server will serve the certificate returned by that function and you can leave the cert file/key empty. (I don't know how it's used further within the http/tls stack.)

I've just started looking at this quic-go, maybe :)

**Comment by marten-seemann** [comment]
> I've just started looking at quic-go, maybe :)

Let me know if you need any pointers on where / how to start.

**Comment by marten-seemann** [comment]
@sotodel Any updates? I'm cutting a new quic-go release (that supports QUIC draft-28) later this week. Would nice to get this change in.

**Comment by sotodel** [comment]
@marten-seemann Unfortunately, no. I won't have anything ready anytime soon.

I gave it a very brief look over and I _think_ the change is relatively simple. The method would just look like its `http.Server` counterpart (really in `ServeTLS`)
 
```
 func (s *Server) ListenAndServeTLS(certFile, keyFile string) error {
       config := s.TLSConfig // not sure if this needs to be reset
       if config == nil {
               config = &tls.Config{}
       }       else {
               config = config.Clone()
       }

       configHasCert := len(config.Certificates) > 0 || config.GetCertificate != nil
       if !configHasCert || certFile != "" || keyFile != "" {
               var err error
               config.Certificates = make([]tls.Certificate, 1)
               config.Certificates[0], err = tls.LoadX509KeyPair(certFile, keyFile)
               if err != nil {
                       return err
               }
        }
        return s.serveImpl(config, nil)
}
```

However, when I tried to set up a sample (quic-go) server with a random self-signed certificate, the sample client (also quic-go) did not properly verify/reject it. I don't know if I configured all that correctly, it was done kind of in a rush.

I can try again at some point next week.

**Comment by marten-seemann** [comment]
> I can try again at some point next week.

Sounds good. We'll get it in the next release then.

**Comment by marten-seemann** [comment]
I'm confused by this code snippet: https://github.com/golang/go/blob/0eb9767caad4c4a0b447824d7b27328864612d58/src/net/http/server.go#L2969-L2977

It doesn't look like `config.GetCertificate` is ever called. The code in the `if` just loads the certificate from `certFile` (even if both `certFile` and `keyFile` are not set.

**Comment by sotodel** [comment]
Right, it's expected to be set before calling the method
```
// Files containing a certificate and matching private key for the
// server must be provided if neither the Server's
// TLSConfig.Certificates nor TLSConfig.GetCertificate are populated.
```
During the [server TLS handshake](https://github.com/golang/go/blob/0eb9767caad4c4a0b447824d7b27328864612d58/src/crypto/tls/handshake_server.go#L222), this unexported [`getCertificate` method](https://golang.org/src/crypto/tls/common.go?s=16110:25304#L867) is called to select a certificate by calling `tls.GetCertificate` or by finding an appropriate one in `tls.Certificates`.

`tls.GetCertificate` is useful if your server has custom rules for serving certificates based on [`ClientHelloInfo`](https://golang.org/pkg/crypto/tls/#ClientHelloInfo) information like the SNI or the signature schemes.

**Comment by sotodel** [comment]
> However, when I tried to set up a sample (quic-go) server with a random self-signed certificate, the sample client (also quic-go) did not properly verify/reject it. I don't know if I configured all that correctly, it was done kind of in a rush.

Yeah I was definitely doing things incorrectly. I figured it out last night and it works nicely. I believe the `ListenAndServeTLS` snippet in my comment [here](https://github.com/lucas-clemente/quic-go/issues/2514#issuecomment-634057777) works. I'm using a test server like

```
package main

import (
	"crypto/tls"
	"errors"
	"flag"
	"log"
	"net/http"
	"os"
	"os/signal"
	"path/filepath"
	"strings"
	"sync"
	"syscall"
	"github.com/lucas-clemente/quic-go"
	"github.com/lucas-clemente/quic-go/http3"
)

func main() {
	listenAddress := flag.String("listen", ":4443", "Listen address. Default: ':4443'")
	flag.Parse()

	domainNamesToCertificates := make(map[string]tls.Certificate)
	for _, fileLocation := range flag.Args() {
		domainName := filepath.Base(fileLocation)
		certificate, err := tls.LoadX509KeyPair(fileLocation, fileLocation)
		if err != nil {
			log.Fatalf("Failed to parse certificate from '%s': %s", fileLocation, err)
		}
		domainNamesToCertificates[domainName] = certificate
	}
	getCertificateFromSNI := func(clientHello *tls.ClientHelloInfo) (*tls.Certificate, error) {
		if certificate, ok := domainNamesToCertificates[clientHello.ServerName]; ok {
			return &certificate, nil
		}
		return nil, errors.New("no certificate matching the given server name")
	}
	tlsConfig := &tls.Config{
		GetCertificate: getCertificateFromSNI,
	}
	mux := http.NewServeMux()
	mux.HandleFunc("/", func(w http.ResponseWriter, req *http.Request) {
		w.WriteHeader(http.StatusNoContent)
	})
	quicConf := &quic.Config{}
	quicServer := http3.Server{
		Server:     &http.Server{TLSConfig: tlsConfig, Handler: mux, Addr: *listenAddress},
		QuicConfig: quicConf,
	}

	exitDone := &sync.WaitGroup{}
	exitDone.Add(1)

	go func() {
		sigint := make(chan os.Signal, 1)
		signal.Notify(sigint, syscall.SIGINT, syscall.SIGQUIT)
		_ = <-sigint

		if err := quicServer.Close(); err != nil {
			log.Printf("Failed to shutdown QUIC HTTP server: %v", err)
		}
		exitDone.Done()
	}()
	if err := quicServer.ListenAndServeTLS("", ""); err != http.ErrServerClosed {
		if !strings.Contains(err.Error(), "server closed") {
			log.Fatal("Error starting QUIC HTTP server: ", err)
		}
	}
	exitDone.Wait()
}
```
With the two empty strings passed and because `tls.Config.GetCertificate` is set, the quic http3 Server will defer certificate selection to that `getCertificateFromSNI` function (which uses SNI).

However, I hadn't actually noticed the `ListenAndServe` method in http3/server.go. I overlooked it because of the name missing TLS. A method with the same name in `http.Server` doesn't configure TLS so that had me confused (it's not possible afaik to use HTTP3 without TLS). The method `http3.Server.ListenAndServer` method doesn't do anything to the underlying `tls.Config`, it uses it as given. In that sense, if `GetCertificate` is set, it'll use it.

You don't really need the `ListenAndServerTLS` version I'm proposing, unless you want to keep some parity with the standard `http.Server` API.

**Comment by astrolox** [comment]
FYI: This is needed in order to use the autocert module to automatically provision LetsEncrypt certificates.
https://pkg.go.dev/golang.org/x/crypto/acme/autocert

**Comment by watjurk** [comment]
Any update on this?

**Comment by marten-seemann** [comment]
I think this has been resolved a long time ago, and Caddy is actually using it.
